src/components/CreditCardForm.js

```diff
@@ -0,0 +1,141 @@
+import React, { useState } from "react";
+
+function CreditCardForm(props) {
+  const userDefault = {
+    firstName: {
+      value: "",
+    },
+    surname: {
+      value: "",
+    },
+    cardNumber: {
+      value: "",
+      error: "",
+    },
+    expirationDate: {
+      value: "",
+    },
+  };
+
+  const [user, setUser] = useState(userDefault);
+  const [creditCardType, setCreditCardType] = useState("");
+
+  function checkCardNumber(value) {
+    if (value.length >= 13 && value.length <= 16) {
+      let newNum = value.padStart(16, "0");
+      let sum = 0;
+      let newSum = 0;
+      for (let i = 0; i < newNum.length; i++) {
+        if (i === 0) {
+          sum = (parseInt(newNum[i]) * 2).toString();
+        } else if (i % 2 === 0) {
+          sum = sum + (parseInt(newNum[i]) * 2).toString();
+        } else {
+          sum = sum + (parseInt(newNum[i]) * 1).toString();
+        }
+      }
+      for (let j = 0; j < sum.length; j++) {
+        newSum = newSum + parseInt(sum[j]);
+      }
+      let newSumStr = newSum.toString();
+      if (newSumStr.charAt(newSumStr.length - 1) !== "0") {
+        throw new Error("Wrong card number!");
+      } else {
+        console.log("Correct number!");
+      }
+    }
+  }
+
+  function handleChange(e) {
+    const { name: field, value } = e.target;
+    let error = "";
+    if (typeof user[field] !== "undefined") {
+      try {
+        checkCardNumber(value);
+      } catch (err) {
+        error = err.message;
+      }
+      setUser({ ...user, [field]: { value, error } });
+    }
+    setCreditCardType("");
+  }
+
+  function CreditCardTypeCheck() {
+    if (cardNumber.value.length === 16 && cardNumber.value.charAt(0) === "5") {
+      setCreditCardType("MasterCard");
+    } else if (
+      cardNumber.value.length === 15 &&
+      cardNumber.value.charAt(0) === "3"
+    ) {
+      setCreditCardType("AmericanExpress");
+    } else if (
+      cardNumber.value.charAt(0) === "4" &&
+      (cardNumber.value.length === 13 || cardNumber.value.length === 16)
+    ) {
+      setCreditCardType("Visa");
+    } else {
+      setCreditCardType("Card number do not match any type!");
```

Comment:
Przydałoby się wymyśleć rozwiązanie bardziej uniwersalne - w takiej wersji dodanie innych 50 kart będzie bardzo uciążliwe. Jak zawsze może pomóc odpowiedna struktura danych np.
```
creditCardRules = {
  ''Visa": { // nazwa producent
     length: [13, 16], // możliwe długości
     chars: {
       0: "4"  // pozycja => znak
     }
  },
  ''MasterCard": {
     length: [15],
     chars: {
       0: "3" 
     }
  }
}
```
PS. To tak na przyszłość - obecnie już bym nie zmieniał ;)